Documentation/teaching/labs/kernel_profiling.rst

```diff
@@ -0,0 +1,473 @@
+
+We see that the sum of the ``user`` and ``sys`` values doesn't add up to the
+``real`` value. This happens either when the application runs on multiple cores,
+in which case the sum might be higher, or the application sleeps, in which case
+the sum is lower.
+
+top
+---
+
+``top`` is an application that is found on most systems which lists in real time
+the applications that are running on the system. ``top`` runs interactively, and
+it auto-refreshes its output, as opposed to ``ps``. We use this tool when we
+want a high level of continuous monitoring.
+
+Profiling Methodology
+=====================
+
+When doing profiling, our goal is to identify the cause of a problem. Usually
+this problem is observed by someone when their application doesn't work as
+expected. When we say that an application did not work as expected, this can
+mean different things for different people. For example, one person might
+complain that the application has a slowdown, while another might say that the
+application runs on the CPU, but it doesn't output anything.
+
+The first step in any problem solving context is to understand the default
+behaviour of the application we're trying to debug, and to make sure that it is
+now not running in the expected parameters.
+
+Exercises
+=========
+
+.. include:: ../labs/exercises-summary.hrst
+.. |LAB_NAME| replace:: kernel_profiling
+
+.. note::
+
+    This session will require us to use the ``perf`` tracing tool. When running
+    natively on our systems, we have to install the
+    ``linux-tools-<version>-generic`` package using a package manager in order
+    to run it. Because in our visual machine we don't have access to a package
+    manager, we will be downloading the ``perf`` binary from `this
+    <http://swarm.cs.pub.ro/~sweisz/perf>`_ link. Download the application in
+    the ``skels/kernel_profiling`` directory, and grant in execution
+    permissions.
+
+.. warning::
+
+    When running ``perf``, make sure that you're running the downloaded version,
+    not the version in the ``PATH`` variable.
+
+.. note::
+
+    When going through this session's exercises, we will have to run command in
+    parallel. In order to do this, we will have to connect to the virtual machine
+    using SSH. We recommend using the ``core-image-sato-sdk-qemu`` image, since it
+    has the tools that we need. To run the virtual machine using the
+    ``core-image-sato-sdk-qemu`` file system, uncomment line 16 in the
+    ``qemu/Makefile`` file.
+
+.. note::
+
+    If you wish to run the ``perf-tools`` based scripts that we have included in
+    the repository, such as ``iosnoop.sh``, you will have to grant it execution
+    privilleges, in order to be copied to the virtual machine file system.
+
+.. note::
+
+    In order to improve the course of SO2, its components and the way it is
+    conducted, your opinions are very useful to us.
+    `curs.upb.ro platform <https://curs.upb.ro/mod/feedbackadm/view.php?id=26596>`_
```

Comment:
Please update link to: https://curs.upb.ro/2021/mod/feedbackadm/view.php?id=16907